### PR TITLE
Stdlib: Add take, drop, take_while and drop_while to the List module

### PR DESCRIPTION
**Pull request by kit-ty-kate**
https://github.com/ocaml/ocaml/pull/9968 but without the controversial `split_at`.

I've also updated the behaviour of `take` and `drop` to match the behaviour of `Seq.take` and `Seq.drop`, as well as taken into account the reviews from @nojb and @craigfe (https://github.com/ocaml/ocaml/pull/9968#pullrequestreview-506146952)

cc @0scarB @gasche

### CODE REVIEW

**Review by gasche — APPROVED** [review]
I believe that this is correct.

The implementation of `take{,_while}` uses acc-and-reverse rather than tail_mod_cons. I don't know which approach is preferable (the `tail_mod_cons` is faster and the code looks simpler, but it is more "magic" and compiles into more code), so I am happy to follow the choice of the author.

**Comment by gasche** [comment]
Also: I reviewed the .mli and compared to seq.mli for the same functions. The wording is slightly different in places, and I am not sure which one is better. I would have had a slight preference for using exactly the same wording for consistency, but I decided not to both as the differences are small.

**Comment by gasche** [comment]
As a stdlib PR, this would need a second approval from a maintainer before we could consider merging it.

**Comment by 0scarB** [comment]
Awesome, thanks @kit-ty-kate! I'm also interested in whether `tail_mod_cons` is faster / less memory intensive as @gasche mentioned but don't know enough about FP to make an  off-the-cuff judgement. I'll try to do some newbie profiling and report back.

**Comment by 0scarB** [comment]
I looked at the implementations that Containers and Base use for `take`/`drop`, modified them to raise the `Invalid_argument` exception and benchmarked a bit https://github.com/0scarB/ocaml-list-take-drop-bench/blob/7f8bc5e456207879eece16b595e418f0aed29e60/bench/bench.ml#L1-L134. @gasche you're correct, implementations of `take` with `tail_mod_cons` need less heap allocations **for large lists**: 
image
> Change in `Gc.stat.major_heap + Gc.stat.minor_heap` before and after calls to `take (n/2) <list of chars of size n>` implementations, for `0 <= n <= 100_000`, using a rolling average with a window of size 10.

This also makes implementations with `tail_mod_cons` slightly faster -- presumably due to few GC pauses:
image
> Change in `Sys.time` in seconds before and after calls to `take (n/2) <list of chars of size n>` implementations, for `0 <= n <= 100_000`, using a rolling average with a window of size 10.

`drop` on the other hand does not grow the heap as expected:
image
Resulting evenly paced, linear runtimes across all implementations (the bottom left of the plot does look a bit weird though):
image

**Comment by 0scarB** [comment]
Haven't benchmarked the `_while` variants yet but the same should hold. I.e. `take_while` can be made slightly more performant with `tail_mod_cons`.

File: stdlib/list.ml
Comment (0scarB) [inline stdlib/list.ml]:
```ocaml
  let take n l =
    if n < 0 then invalid_arg "List.take"
    else
      let[@tail_mod_cons] rec loop n l =
        if n = 0 then []
        else
          match l with
          | [] -> []
          | x::l -> x::loop (n - 1) l
      in loop n l
```

One option for a slightly more performant version with `@tail_mod_cons`.

**Comment by kit-ty-kate** [comment]
done

File: stdlib/list.ml
Comment (kit-ty-kate) [reply stdlib/list.ml]: I used a slightly more concise version instead as i felt this version could be improved visually a little (less vertical space), could you check that it behave the same performance wise? I also changed `take_while` to use tail_mod_cons.

Comment (0scarB) [reply stdlib/list.ml]: Yep the new version seems to have the same runtime and heap allocation behaviour. Very elegant way to reduce the vertical space, nice work! 
image
image

**Review by 0scarB — APPROVED** [review]

File: stdlib/list.mli
Comment (nojb) [inline stdlib/list.mli]:
```suggestion
    or [l] itself if [n > length l].
```
This is not, strictly speaking, true: it is not `l` that is returned, but a _copy_ of `l`. The implementation could be tweaked to make the comment true, though.

**Review by nojb — COMMENTED** [review]
LGTM, modulo a small point, see below.

File: stdlib/list.mli
Comment (kit-ty-kate) [reply stdlib/list.mli]:
```suggestion
    or a copy of [l] if [n > length l].
```
Let’s fix the documentation then. I’m not sure it’s worth the niche optimization. If someone thinks it is, I can revert and change the code.

File: stdlib/list.mli
Comment (ferminr) [inline stdlib/list.mli:368]: `or a copy of [l] if [n >= length l]`
(similarly for drop)

Comment (kit-ty-kate) [inline stdlib/list.mli:368]:
```suggestion
    or a copy of [l] if [n >= length l].
```

File: stdlib/list.mli
Comment (kit-ty-kate) [inline stdlib/list.mli:378]:
```suggestion
    or [[]] if [n >= length l].
```

File: stdlib/listLabels.mli
Comment (kit-ty-kate) [inline stdlib/listLabels.mli]:
```suggestion
    or a copy of [l] if [n >= length l].
```

File: stdlib/listLabels.mli
Comment (kit-ty-kate) [inline stdlib/listLabels.mli:378]:
```suggestion
    or [[]] if [n >= length l].
```

File: stdlib/list.mli
Comment (kit-ty-kate) [reply stdlib/list.mli:368]: done

Comment (yallop) [reply stdlib/list.mli:368]: The versions with `>` are better.  The special case is when `length (take n l) <> n`, i.e. when `n > length l`.

File: stdlib/list.mli
Comment (kit-ty-kate) [inline stdlib/list.mli]:
```suggestion
    or a copy of [l] if [n > length l].
```

File: stdlib/list.mli
Comment (kit-ty-kate) [inline stdlib/list.mli]:
```suggestion
    or [[]] if [n > length l].
```

File: stdlib/listLabels.mli
Comment (kit-ty-kate) [inline stdlib/listLabels.mli]:
```suggestion
    or [[]] if [n > length l].
```

File: stdlib/listLabels.mli
Comment (kit-ty-kate) [inline stdlib/listLabels.mli]:
```suggestion
    or a copy of [l] if [n > length l].
```

File: stdlib/list.mli
Comment (kit-ty-kate) [reply stdlib/list.mli:368]: reverted

**Review by nojb — APPROVED** [review]
LGTM

**Comment by nojb** [comment]
@gasche there were a few changes after your approval. Are you happy with the current state of the PR?

File: Changes
Comment (gasche) [inline Changes]: (It's a minor point but I would move Oscar to co-author.)

**Review by gasche — APPROVED** [review]
I'm still happy with the change. Let's leave @kit-ty-kate the opportunity to cleanup the history if they want to, and then merge.

File: Changes
Comment (0scarB) [reply Changes]: It's appreciated. I don't mind either way. Whatever's easiest for you @kit-ty-kate :)

Comment (kit-ty-kate) [reply Changes]: done

Comment (gasche) [reply Changes]: Not quite yet!

```
Checking c4bce87a16c60d8a128787929f0c977f317ba4d4: Changes
./Changes:15.81: [long-line] line is over 80 columns
```

Comment (kit-ty-kate) [reply Changes]: fixed.

**Comment by dbuenzli** [comment]
Is there any reason why this was made to raise `Invalid_argument` if `n` is negative ? 

This is highly non standard in my eyes. Neither Haskell, nor `containers`, nor `batteries`, nor `base` do so, they all return `[]` in this case.

The functions have been in only since 5.3 and you are not supposed to trigger `Invalid_argument` could we perhaps still change this ?

**Comment by nojb** [comment]
Thanks @dbuenzli for raising this point. I agree that having these functions be total is generally nicer.

> The functions have been in only since 5.3 and you are not supposed to trigger `Invalid_argument` could we perhaps still change this ?

I would be in favour of doing this, indeed.

**Comment by gasche** [comment]
I'm also fine with it. (If people want to support 5.3 they will still need to be defensive about this in their code.)

**Comment by dbuenzli** [comment]
Cool I'll do a PR.

Just FTR, the reason why I'm asking is that I'm contemplating adding the counter part of these functions in `String`. 

All the various string extension libraries or ad-hoc reimplementation of these functions I have been using do not raise on negative `n`. So I'd rather not have that behaviour in `String` but I could already forsee the discussion about "Ah but `List.{take,drop}` do raise on negative `n`".

**Pull request merged**
